File: week-6/py/member_system/repository/mysql_repository/mysql_unit_of_work.py
import json
import mysql.connector
from mysql.connector.pooling import MySQLConnectionPool
from member_system.repository.mysql_repository.mysql_member_repository import MySQLMemberRepository
from member_system.repository.mysql_repository.mysql_message_repository import MySQLMessageRepository
from member_system.repository.unit_of_work import MemberRepository, MessageRepository, UnitOfWork

class MySQLUnitOfWork(UnitOfWork):
    def __init__(self, configPath: str, debug=False):
        self.__debug = debug
        config = self.loadConfig(configPath)
        self.initDb(config)
        self.__cnxpool = MySQLConnectionPool(pool_name='mypool', pool_size=4, **config)
        UnitOfWork.__init__(self)

    def __del__(self):
        if self.__debug:
            self.messageRepository.dropTableIfExists()
            self.memberRepository.dropTableIfExists()

    def _createMemberRepository(self) -> MemberRepository:
        return MySQLMemberRepository(self.__cnxpool, self.__debug)

    def _createMessageRepository(self) -> MessageRepository:
        return MySQLMessageRepository(self.__cnxpool, self.memberRepository.tableName, self.__debug)

    @staticmethod
    def isAvailable(configPath: str) -> bool:
        try:
            config = MySQLUnitOfWork.loadConfig(configPath)
            mysql.connector.connect(**config)
            return True
        except:
            return False

    @staticmethod
    def loadConfig(configPath: str):
        with open(configPath) as file:
            return json.load(file)
    
    @staticmethod
    def initDb(config):
        database = config['database']
        config['database'] = None
        cnx = mysql.connector.connect(**config)
        cursor = cnx.cursor()
        cursor.execute('SHOW DATABASES;')
        if not (database,) in cursor.fetchall():
            cursor.execute('CREATE DATABASE ' + database)
        cursor.close()
        cnx.close()
        config['database'] = database

File: week-6/go/membersys/repository/mysql_unit_of_work.go
package repository

import (
	. "Project/membersys/core"

	"database/sql"

	_ "github.com/go-sql-driver/mysql"
)

type MySQLUnitOfWork struct {
	UnitOfWork
	config string
	debug  bool
}

func (u *MySQLUnitOfWork) Init() {
	db, _ := sql.Open("mysql", u.config)
	mysql := MySQLRepository{db: db, debug: u.debug}

	memberRepository := new(MySQLMemberRepository)
	memberRepository.MySQLRepository = mysql
	messageRepository := new(MySQLMessageRepository)
	messageRepository.MySQLRepository = mysql
	messageRepository.memberTableName = memberRepository.TableName()

	u.MemberRepository = memberRepository
	u.MessageRepository = messageRepository

	if u.debug {
		messageRepository.DropTable(messageRepository)
		memberRepository.DropTable(memberRepository)
		memberRepository.CreateTable(memberRepository)
		messageRepository.CreateTable(messageRepository)
	}
}

func NewMySQLUnitOfWork(config string, debug bool) *MySQLUnitOfWork {
	unitOfWork := MySQLUnitOfWork{config: config, debug: debug}
	InitUnitOfWork(&unitOfWork)
	return &unitOfWork
}
